### PR TITLE
Upgrade from Webpacker 3 to Shakapacker 7 and React-Rails 3

### PR DESCRIPTION
**Pull request by ahangarha**
This PR upgrades the project as following:
- From Webpacker 3 to Shakapacker 7 (step by step)
- From React-Rails 2 to React-Rails 3 (for SSR support)

We may keep the branch (maybe with a more appropriate name) like other branches in this project.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -24,3 +24,10 @@
 /public/packs
 /public/packs-test
 /node_modules
+
+/public/packs
+/public/packs-test
+/node_modules
```
Comment (Judahmeek) [inline .gitignore]: Why is this duplicated on lines 24-26?

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -6,6 +6,7 @@
 
     <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track': 'reload' %>
     <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
+    <%= javascript_pack_tag "application" %>
```
Comment (Judahmeek) [inline app/views/layouts/application.html.erb:8]: Why is `application` in both the `javascript_include_tag` & `javascript_pack_tag`?

File: config/initializers/assets.rb [hunk 3]
```diff
@@ -12,3 +12,6 @@
 # application.js, application.css, and all non-JS/CSS in the app/assets
 # folder are already added.
 # Rails.application.config.assets.precompile += %w( admin.js admin.css )
+
+# Add node_modules folder to the asset load path.
+Rails.application.config.assets.paths << Rails.root.join('node_modules')
```
Comment (Judahmeek) [inline config/initializers/assets.rb]: Why was this added?

**Review by Judahmeek — CHANGES_REQUESTED** [review]
I've got some questions.

**Comment by justin808** [comment]
@Judahmeek can review and merge.

File: .gitignore (hunk 1 above)
Comment (ahangarha) [reply .gitignore]: It must be coming from some rebase 🤔

**Comment by ahangarha** [comment]
@Judahmeek 
Thanks for the review.

I am wondering why those duplicate lines were added. Since it was a long time since I started working on this PR, I don't remember what I was doing then.

We get some deprecation warnings in the terminal, and I believe it is because this project uses Rails 5. Do you think it should be fixed in this PR, or can it have another PR?

File: .nvmrc [hunk 4]
```diff
@@ -0,0 +1 @@
+14
```
Comment (justin808) [inline .nvmrc]: why old version?

File: Gemfile [hunk 5]
```diff
@@ -5,19 +5,19 @@ git_source(:github) do |repo_name|
   "https://github.com/#{repo_name}.git"
 end
 
-
 # Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
-gem 'rails', '~> 5.1.4'
+gem "rails", "~> 5.2"
```
Comment (justin808) [inline Gemfile:9]: v6?

File: Gemfile [hunk 6]
```diff
@@ -5,19 +5,19 @@ git_source(:github) do |repo_name|
...
 # Use sqlite3 as the database for Active Record
 gem 'sqlite3'
 # Use Puma as the app server
 gem 'puma', '~> 4.3'
 # Use SCSS for stylesheets
 gem 'sass-rails', '~> 5.0'
+# gem 'sassc-rails'
```
Comment (justin808) [inline Gemfile]: remove dead coee?

File: bin/yarn [hunk 7]
```diff
@@ -1,9 +1,16 @@
 #!/usr/bin/env ruby
-VENDOR_PATH = File.expand_path('..', __dir__)
-Dir.chdir(VENDOR_PATH) do
-  begin
-    exec "yarnpkg #{ARGV.join(" ")}"
-  rescue Errno::ENOENT
+
+APP_ROOT = File.expand_path("..", __dir__)
```
Comment (justin808) [inline bin/yarn:3]: why is this file not deleted?

File: config/webpack/clientWebpackConfig.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  // server-bundle is special and should ONLY be built by the serverConfig
+  // In case this entry is not deleted, a very strange "window" not found
+  // error shows referring to window["webpackJsonp"]. That is because the
+  // client config is going to try to load chunks.
+  delete clientConfig.entry['server_rendering']
+  
+  return clientConfig
+}
+
+module.exports = configureClient
```
Comment (justin808) [inline config/webpack/clientWebpackConfig.js]: no eol at eof

File: config/webpack/serverWebpackConfig.js [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+  // If using the default 'web', then libraries like Emotion and loadable-components
+  // break with SSR. The fix is to use a node renderer and change the target.
+  // If using the React on Rails Pro node server renderer, uncomment the next line
+  // serverWebpackConfig.target = 'node'
+
+  return serverWebpackConfig
+}
+
+module.exports = configureServer
```
Comment (justin808) [inline config/webpack/serverWebpackConfig.js]: fix the red dots!

**Review by justin808 — CHANGES_REQUESTED** [review]

File: Gemfile (hunk 5 above)
Comment (ahangarha) [reply Gemfile:9]: All in one PR?
I thought Upgrading Rails could and should have its own PR. And this little change was to make it possible to build the project. I don't remember now but there was an issue with Rails < 5.2. I couldn't run the app.

File: bin/yarn (hunk 7 above)
Comment (ahangarha) [reply bin/yarn:3]: Even in Shakapacker, we still have this file.

File: .nvmrc (hunk 4 above)
Comment (ahangarha) [reply .nvmrc]: I removed the file.

**Pull request merged**